### DISCUSSION TITLE
Should stub functions fail?

### DISCUSSION THREAD

**Discussion by samueljsb** [python/typing#1269] [General]
Currently it's possible to write a function that has a non-`None` return type but returns None, e.g:

```python
def f() -> int: pass
def g() -> str: ...
```

This is considered fine by mypy, but will cause type-related problems at runtime:

```pycon
>>> def f() -> int: pass
...
>>> f() + 1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for +: 'NoneType' and 'int'
```

Why does mypy allow these functions?

**Comment by hauntsaninja** [comment]
On mypy master, we've modified heuristics to complain about this. See:
https://github.com/python/mypy/pull/13729 
https://github.com/python/mypy/pull/12118 
https://github.com/python/mypy/issues/2350
